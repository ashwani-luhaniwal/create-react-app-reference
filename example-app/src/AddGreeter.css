.AddGreeter {
    margin: 20px;
    padding: 20px;
    border: 2px solid #00ffdb;
    background: #ddffee;
    text-align: center;
}