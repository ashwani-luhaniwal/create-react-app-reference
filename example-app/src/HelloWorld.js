import React, { Component } from 'react';
import './HelloWorld.css';

/**
 * props are properties inside of React component that are passed in from somewhere
 * generally from parent.
 * @param {props} param0 
 */
/*const HelloWorld = (props) => {
    return (<div className="HelloWorld">Hello {props.name}</div>);
};
export default HelloWorld;*/

// HelloWorld component extends functionality of Component Class 
class HelloWorld extends Component {
    constructor(props) {
        // calls parent's constructor using super(props).
        super(props);
        // set property on class called "state"
        this.state = { greeting: 'Hello' };
        // Tells JavaScript, any time you see a reference to "this" inside of the frenchify
        // function, please refer to ME.
        this.frenchify = this.frenchify.bind(this);
    }
    frenchify() {
        // setState only changes keys that are specified inside of state, it doesn't replace
        // the entire state.
        this.setState({ greeting: 'Bonjour' });
    }
    render() {
        return (
            <div className="HelloWorld">
                {this.state.greeting} {this.props.name}!
                <br />
                <button onClick={this.frenchify}>Frenchify!</button>
            </div>
        );
    }
}
export default HelloWorld;