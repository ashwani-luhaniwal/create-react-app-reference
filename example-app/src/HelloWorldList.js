import React, {Component} from 'react';
import './HelloWorldList.css';
import HelloWorld from './HelloWorld';
import AddGreeter from './AddGreeter';

export default class HelloWorldList extends Component {
    constructor(props) {
        super(props);
        this.state = { greetings: ['Jim', 'Sally', 'Ashwani'] };
        this.renderGreetings = this.renderGreetings.bind(this);
        this.addGreeting = this.addGreeting.bind(this);
        this.removeGreeting = this.removeGreeting.bind(this);
    }
    renderGreetings() {
        // "map" each item in array to special anonymous function which returns HelloWorld
        // component. The "key" is here because React wants to know which element to modify/
        // remove when one of the elements in your list changes, it has to be able to
        // uniquely identify which element it is, so here we are just specifying key as name.
        return this.state.greetings.map(name => (
            <HelloWorld key={name} name={name} removeGreeting={this.removeGreeting} />
        ));
    }
    addGreeting(newName) {
        // add a new element onto list of greetings
        this.setState({ greetings: [...this.state.greetings, newName] });
    }
    removeGreeting(removeName) {
        // filter our list of greetings only to greetings that do not match the name of 
        // the greeting we want to remove
        const filteredGreetings = this.state.greetings.filter(name => {
            return name !== removeName;
        });
        this.setState({ greetings: filteredGreetings });
    }
    render() {
        return (
            <div className="HelloWorldList">
                <AddGreeter addGreeting={this.addGreeting} />
                {this.renderGreetings()}
            </div>
        );
    }
}