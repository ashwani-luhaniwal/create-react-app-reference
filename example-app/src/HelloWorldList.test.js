import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import HelloWorldList from './HelloWorldList';
import AddGreeter from './AddGreeter';
import HelloWorld from './HelloWorld';

describe(HelloWorldList, () => {
    const component = shallow(
        <HelloWorldList />
    );

    // snapshot test
    it('renders and matches our snapshot', () => {
        const component = renderer.create(
            <HelloWorldList />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    // include AddGreeter component in our component
    it('contains an AddGreeter subcomponent', () => {
        /**
         * We use .find(<element>) to find a DOM element (via selectors) or class name for a 
         * component (such as AddGreeter here) to find specific elements.
         */
        expect(component.find(AddGreeter)).toHaveLength(1);
    });

    // we render the number of "HelloWorld" components that we have entries in greetings state
    // in the component
    it('contains the same number of HelloWorld components as greetings', () => {
        const helloWorlds = component.find(HelloWorld).length;
        // Use .state(<key>) function on a component to fetch a value out of state
        const greetings = component.state('greetings').length;
        // we expect, given how "renderGreetings" works, to find same number of "HelloWorld"
        // components as the number of entries as the "greetings" array in our component's state.
        expect(helloWorlds).toEqual(greetings);
    });

    // test for addGreeting function
    it('adds another greeting when the add greeting function is called', () => {
        const before = component.find(HelloWorld).length;
        /**
         * If you want to call a function directly on a component in a test, you do so by
         * referencing the component, calling the .instance() function on it, and then
         * calling whatever function we want to call from there (such as addGreeting)
         */
        component.instance().addGreeting('Sample');
        // Update the component (this is important as you test will fail without it)
        component.update();
        const after = component.find(HelloWorld).length;
        expect(after).toBeGreaterThan(before);
    });

    // test for removeGreeting function
    it('removes a greeting from the list when the remove greeting function is called', () => {
        const before = component.find(HelloWorld).length;
        // the removeGreeting expects an entry to remove, so we just grab the first element
        // out of state (again, this ensures that nothing in our tests make assumptions
        // about the initial state or hard-coded values in our components)
        const removeMe = component.state('greetings')[0];
        component.instance().removeGreeting(removeMe);
        component.update();
        const after = component.find(HelloWorld).length;
        expect(after).toBeLessThan(before);
    });
});