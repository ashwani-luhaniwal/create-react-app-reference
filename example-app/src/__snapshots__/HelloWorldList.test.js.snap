// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class HelloWorldList extends _react.Component {
    constructor(props) {
        super(props);
        this.state = { greetings: ['Jim', 'Sally', 'Ashwani'] };
        this.renderGreetings = this.renderGreetings.bind(this);
        this.addGreeting = this.addGreeting.bind(this);
        this.removeGreeting = this.removeGreeting.bind(this);
    }
    renderGreetings() {
        // "map" each item in array to special anonymous function which returns HelloWorld
        // component. The "key" is here because React wants to know which element to modify/
        // remove when one of the elements in your list changes, it has to be able to
        // uniquely identify which element it is, so here we are just specifying key as name.
        return this.state.greetings.map(name =>
        _react2.default.createElement(_HelloWorld2.default, { key: name, name: name, removeGreeting: this.removeGreeting, __source: { fileName: _jsxFileName, lineNumber: 20 }, __self: this }));

    }
    addGreeting(newName) {
        // add a new element onto list of greetings
        this.setState({ greetings: [...this.state.greetings, newName] });
    }
    removeGreeting(removeName) {
        // filter our list of greetings only to greetings that do not match the name of 
        // the greeting we want to remove
        const filteredGreetings = this.state.greetings.filter(name => {
            return name !== removeName;
        });
        this.setState({ greetings: filteredGreetings });
    }
    render() {
        return (
            _react2.default.createElement('div', { className: 'HelloWorldList', __source: { fileName: _jsxFileName, lineNumber: 37 }, __self: this },
                _react2.default.createElement(_AddGreeter2.default, { addGreeting: this.addGreeting, __source: { fileName: _jsxFileName, lineNumber: 38 }, __self: this }),
                this.renderGreetings()));


    }} renders and matches our snapshot 1`] = `
<div
  className="HelloWorldList"
>
  <div
    className="AppGreeter"
  >
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
      
    <button
      onClick={[Function]}
    >
      Add
    </button>
  </div>
  <div
    className="HelloWorld"
  >
    Hello
     
    Jim
    !
    <br />
    <button
      className="frenchify"
      onClick={[Function]}
    >
      Frenchify!
    </button>
    <br />
    <button
      className="remove"
      onClick={[Function]}
    >
      Remove Me!
    </button>
  </div>
  <div
    className="HelloWorld"
  >
    Hello
     
    Sally
    !
    <br />
    <button
      className="frenchify"
      onClick={[Function]}
    >
      Frenchify!
    </button>
    <br />
    <button
      className="remove"
      onClick={[Function]}
    >
      Remove Me!
    </button>
  </div>
  <div
    className="HelloWorld"
  >
    Hello
     
    Ashwani
    !
    <br />
    <button
      className="frenchify"
      onClick={[Function]}
    >
      Frenchify!
    </button>
    <br />
    <button
      className="remove"
      onClick={[Function]}
    >
      Remove Me!
    </button>
  </div>
</div>
`;
